### PR TITLE
Explicit spatial transformer

### PR DESCRIPTION
**Pull request by mcabbott**
This previously didn't have a manifest

### CODE REVIEW

File: vision/spatial_transformer/spatial_transformer.jl [hunk 1]
```diff
@@ -174,16 +174,16 @@ classifier =
...
 
-ps = Flux.params(localization_net, classifier)
+# ps = Flux.params(localization_net, classifier)
 ## ====
 # create sampling grid
 const sampling_grid = get_sampling_grid(args[:img_size]...) |> dev
 ## ====
 
-opt = Adam(1e-4)
+opt = Flux.setup(Adam(1f-4), (localization_net, classifier))
```
Comment (mcabbott) [inline vision/spatial_transformer/spatial_transformer.jl:183]: Not sure if this is too surprising. Rather than make a state tree per model, it makes one for the Tuple, and then calls `Flux.update!(opt, (localization_net, classifier), grads)` on both in one go.

Comment (ToucheSir) [reply vision/spatial_transformer/spatial_transformer.jl:183]: To reduce the surprise, we could destructure `grads` as well and re-tuple for the `update!` here. Slightly more verbose but more explicit.

Comment (mcabbott) [reply vision/spatial_transformer/spatial_transformer.jl:183]: In the VAE I just made two optimisers etc, like so:

https://github.com/FluxML/model-zoo/pull/375/files#diff-d2fcc7489d82937b5b466b70d629d200f23003b00a0e923cfa5031eedfa54d4cR112-R113

Maybe this wants a comment saying you could also do that?

File: vision/spatial_transformer/spatial_transformer.jl [hunk 2]
```diff
@@ -174,16 +174,16 @@ classifier =
         Dense(256, 10),
     ) |> dev
 
-ps = Flux.params(localization_net, classifier)
+# ps = Flux.params(localization_net, classifier)
```
Comment (mcabbott) [inline vision/spatial_transformer/spatial_transformer.jl:177]:
```suggestion
```

File: vision/spatial_transformer/spatial_transformer.jl [hunk 3]
```diff
@@ -79,28 +79,28 @@ end
...
     progress_tracker = Progress(length(train_loader), 1, "Training epoch $epoch :)")
     losses = zeros(length(train_loader))
     for (i, (x, y)) in enumerate(train_loader)
-        loss, grad = withgradient(ps) do
-            model_loss(x, y)
+        loss, grads = withgradient(localization_net, classifier) do ln, cl
+            model_loss(localization_net, classifier, x, y)
         end
-        Flux.update!(opt, ps, grad)
+        Flux.update!(opt, (localization_net, classifier), grads)
```
Comment (mcabbott) [inline vision/spatial_transformer/spatial_transformer.jl:103]:
```suggestion
        # Both the optimisers state `opt` and the gradients match a tuple of the two networks: 
        Flux.update!(opt, (localization_net, classifier), grads)
```

File: vision/spatial_transformer/spatial_transformer.jl (hunk 1 above)
Comment (ToucheSir) [reply vision/spatial_transformer/spatial_transformer.jl:184]: The [PyTorch version](https://pytorch.org/tutorials/intermediate/spatial_transformer_tutorial.html) uses a single net, so if anything it seems like we should bundle them together more tightly. Maybe a custom layer?

Comment (mcabbott) [reply vision/spatial_transformer/spatial_transformer.jl:184]: I'm going to call that out of scope for this PR though. Just adding explicit gradients, without changing too much. Maybe it's good enough...
